dbt/adapters/trino/connections.py

```diff
@@ -64,7 +62,7 @@ def __init__(self, handle):
         self._fetch_result = None
 
     def cursor(self):
-        self._cursor = self.handle.cursor()
+        self._cursor = self.handle.cursor(experimental_python_types=True)
```

Comment:
We should give the ability to the user to opt in/out of the "experimental" feature.
Should we introduce a profile configuration setting?

Reply:
IMHO the conversions in `ConnectionWrapper._escape_value(cls, value)` are all implemented by the `trino-python-client`, which is the right place to implement them. The covered data types are not really that 'experimental', in the sense that the same Python data type constraints apply as in the earlier code. So I consider this some kind of technical debt that we should remove.

Not setting this parameter consistently would also prevent us from consistently fixing #28.

Reply:
This going to be an impactful change for the users, but I agree that we need to introduce it since some Python types were mapped correctly "by accident". 

Perhaps enable `experimental_python_types` by default and introduce a profile configuration setting in case someone would like to use the old approach (to not impact current pipelines and migrate them later on). But I doubt it is a common situation. 

Could you please point out the code in trino-python-client which converts for instance: 
```
if value is None:
    return "NULL"
```

Reply:
Here is the code that handles the prepared statements parameters

https://github.com/trinodb/trino-python-client/blob/e21b06d2a2c4e6a2b95fbab96d9168ae15b28cc9/trino/dbapi.py#L340-L408

As you can see it handles following cases:

None -> NULL (same)
string -> quoted (same)
pyhon datetime -> trino TIMESTAMP (agate doesn't do anything with timezones, also currently the python-trino-client only supports up to [millisecond precision](https://github.com/trinodb/trino-python-client/issues/42), not completely sure why we strip out the precision within dbt-trino) (same)
python date -> trino DATE (same)
NUMBER -> specific handling for integer, decimal (slightly different)

The number handling seems slightly different but without any impact. Basically every number becomes a Decimal through the agate type detection. See [agate docs](https://agate.readthedocs.io/en/latest/api/data_types.html#agate.Number).

<img width="1221" alt="image" src="https://user-images.githubusercontent.com/9265503/163789508-da1d88ae-3510-4cb9-b011-5e24e5a81296.png">

Trino can handle conversions from decimal to other numeric types. Any loss in precision is the result of conversion to the chosen target column type.

```
CREATE TABLE test_delta.test.test_table1 (a integer);
insert into test_delta.test.test_table1 (a) values (decimal '1.2')
select * from test_delta.test.test_table1
```

Reply:
Let's add fallback logic to the "classic" binding manner and eventually remove it (add an extra issue for follow up).

Even if the changes are mostly about `dbt seed` we should however ensure that the users have the option to fallback to the classic functionality with the newer version of dbt-trino in case they will deal with unforeseen exceptions.

Reply:
Note that we have no telemetry functionality, so we would never know if this fallback logic is used or not and would not be able to remove it. 

I played a bit with a 15K records CSV file. Using the standard batch size (1K) it inserts without issues using the prepared statements. If I increase the batch size to 20K, it starts failing because of `LineTooLong` error in Python `http.client`.

When I increase `http.client._MAXLINE` the query succeeds.

IMHO the only risk with this change is handling of big batches. We should definitely make this `http.client._MAXLINE` configurable and document it.

Reply:
How `http.client._MAXLINE` could be configured in dbt-trino ?

Reply:
Yes, I agree, let's set up it in trino-python-client.

Reply:
I think the use cases of `dbt seed` are definitely much broader then test purposes only.

The issue is that some users may have some seed files that may fail if we remove the http client patch. It won't be obvious for them to find out what they need to do. That was my motivation to add this as a default.